ca/server.go

```diff
@@ -808,3 +826,162 @@ func isFinalState(status api.IssuanceStatus) bool {
 
 	return false
 }
+
+func (s *Server) finishRootRotation(expectedRootCA *api.RootCA) error {
+	// when the nodes are all converged on having TLS certificates issued by the correct entity, we can
+	// try to finish the root rotation
+	clusterID := s.securityConfig.ClientTLSCreds.Organization()
+	return s.store.Update(func(tx store.Tx) error {
+		cluster := store.GetCluster(tx, clusterID)
+		if cluster == nil {
+			return fmt.Errorf("unable to get cluster %s", clusterID)
+		}
+
+		// If the RootCA object has changed (because another root rotation was started or because some other node
+		// had finished the root rotation), we cannot finish the root rotation that we were working on.
+		if !equality.RootCAEqualStable(expectedRootCA, &cluster.RootCA) {
+			return errors.New("target root rotation has changed - unable to complete previous root rotation")
+		}
+
+		var signerCert []byte
+		if len(cluster.RootCA.RootRotation.CAKey) > 0 {
```

Comment:
Another possibility is that there is a root rotation, but the certs in the root rotation have changed (because someone wants to rotate to yet a different cert).  So for each cycle of the root rotation reconciliation, we store the `(api).RootCA` value we're trying to converge to, and ensure that all the nodes are converged on that particular one.

When we call `finishRootRotation`, we pass that saved `RootCA` value.  If the cluster's `RootCA` value has changed in the meanwhile, we abort the update (and hence abort finishing the root rotation), because either someone else set `RootRotation` to nil`, or they changed the values for `RootRotation`, so we want to wait for the next cycle of reconciliation.

Since that stored `RootCA` that we're trying to make sure we converge on has been checked to make sure it does have a `RootRotation` value, if it and the cluster's `RootCA` value are equivalent modulo join tokens (which must be true before we proceed with the update), then the cluster's `RootCA` must have the same non-nil `RootRotation` .